### PR TITLE
Add get_live_services_data endpoint and related query

### PR DESCRIPTION
**Pull request by CrystalPea**
Adds get_live_services_data endpoint to services rest and a related query to services_dao.

The query gets data about live services from multiple tables (Service, User, Organisation, FactBilling) and then serialises it so it can be pulled by admin app to create a csv report for our platform admins.

Needed for this story: https://www.pivotaltracker.com/story/show/164829877

Needs to be deployed before this PR: https://github.com/alphagov/notifications-admin/pull/2934

### CODE REVIEW

File: app/dao/services_dao.py [hunk 1]
```diff
@@ -72,6 +75,89 @@ def dao_count_live_services():
...
+        case([
+            (this_year_ft_billing.c.notification_type == 'email', func.sum(this_year_ft_billing.c.notifications_sent))
+        ], else_=0).label("email_totals"),
+        case([
+            (this_year_ft_billing.c.notification_type == 'sms', func.sum(this_year_ft_billing.c.notifications_sent))
+        ], else_=0).label("sms_totals"),
+        case([
+            (this_year_ft_billing.c.notification_type == 'letter', func.sum(this_year_ft_billing.c.notifications_sent))
+        ], else_=0).label("letter_totals"),
+    ).outerjoin(
```
Comment (servingUpAces) [inline app/dao/services_dao.py]: Do we need to filter the services for live (`count_as_live` = True)?

File: tests/app/dao/test_services_dao.py [hunk 2]
```diff
@@ -380,6 +384,34 @@ def test_get_all_user_services_should_return_empty_list_if_no_services_for_user(
...
+    org = create_organisation()
+    service = create_service(go_live_user=sample_user)
+    template = create_template(service=service)
+    create_service(service_name='second', go_live_user=sample_user)
+    template2 = create_template(service=service, template_type='email')
+    dao_add_service_to_organisation(service=service, organisation_id=org.id)
+    create_ft_billing(bst_date='2019-04-20', notification_type='sms', template=template, service=service)
+    create_ft_billing(bst_date='2019-04-21', notification_type='sms', template=template, service=service)
+    create_ft_billing(bst_date='2018-04-20', notification_type='sms', template=template, service=service)
+    create_ft_billing(bst_date='2019-04-20', notification_type='email', template=template2,
```
Comment (servingUpAces) [inline tests/app/dao/test_services_dao.py]: Should expand this test to have more than one service in ft_billing and letters

File: app/dao/services_dao.py [hunk 3]
```diff
@@ -72,6 +75,89 @@ def dao_count_live_services():
...
+        Service.organisation
```
Comment (servingUpAces) [inline app/dao/services_dao.py]: Do we need to filter by `bst_date`?

File: app/dao/services_dao.py (hunk 1 above)
Comment (CrystalPea) [reply app/dao/services_dao.py]: yeah I think so, at least it is part of the story ticket.

File: app/dao/services_dao.py (hunk 3 above)
Comment (CrystalPea) [reply app/dao/services_dao.py]: We need to get totals for current financial year, so I think yes :)

File: tests/app/dao/test_services_dao.py (hunk 2 above)
Comment (CrystalPea) [reply tests/app/dao/test_services_dao.py]: ok 👍

Comment (CrystalPea) [reply tests/app/dao/test_services_dao.py]: I added more data to the test now, so we have:
- one service with all kinds of records in ft_billing
- one service with letter records in ft_billing
- one service without ft billing records or go_live_user to see that it still appears

File: app/dao/services_dao.py (hunk 1 above)
Comment (CrystalPea) [reply app/dao/services_dao.py]: ok I added that as a filter and added a related check to the test ☺️

File: app/dao/services_dao.py [hunk 4]
```diff
@@ -72,6 +75,93 @@ def dao_count_live_services():
     ).count()
 
 
+def dao_fetch_live_services_data():
+    year_start_date, year_end_date = get_current_financial_year()
+    this_year_ft_billing = FactBilling.query.filter(
```
Comment (servingUpAces) [inline app/dao/services_dao.py]: This can just be a filter on the query rather than a subquery

File: tests/app/dao/test_services_dao.py [hunk 5]
```diff
@@ -380,6 +384,44 @@ def test_get_all_user_services_should_return_empty_list_if_no_services_for_user(
     assert len(dao_fetch_all_services_by_user(user.id)) == 0
 
 
+@freeze_time('2019-04-23T10:00:00')
```
Comment (servingUpAces) [inline tests/app/dao/test_services_dao.py]: We could have a test to ensure the date filter is working.

File: tests/app/dao/test_services_dao.py [hunk 6]
```diff
@@ -380,6 +384,44 @@ def test_get_all_user_services_should_return_empty_list_if_no_services_for_user(
...
+    service_2 = create_service(service_name='second', go_live_user=sample_user)
+    create_service(service_name='third')
+    create_service(service_name='not_live', count_as_live=False)
+    template2 = create_template(service=service, template_type='email')
+    template_letter_1 = create_template(service=service, template_type='letter')
+    template_letter_2 = create_template(service=service_2, template_type='letter')
+    dao_add_service_to_organisation(service=service, organisation_id=org.id)
+    create_ft_billing(bst_date='2019-04-20', notification_type='sms', template=template, service=service)
+    create_ft_billing(bst_date='2019-04-21', notification_type='sms', template=template, service=service)
+    create_ft_billing(bst_date='2018-04-20', notification_type='sms', template=template, service=service)
```
Comment (CrystalPea) [inline tests/app/dao/test_services_dao.py]: @servingUpAces this ft_billing is dated at 2018 so it does not appear in the final result, is this sufficient as a test?

File: tests/app/dao/test_services_dao.py (hunk 5 above)
Comment (CrystalPea) [reply tests/app/dao/test_services_dao.py]: one of ft_biling instances in the test is dated at 2018 and does not appear in the results, so it kind of is tested, or do you mean like a separate test?

File: app/dao/services_dao.py (hunk 4 above)
Comment (CrystalPea) [reply app/dao/services_dao.py]: but then services without ft_billing records in the dates would not show up in the results at all, right? Or at least this is what happened when I tried to apply it as a filter first...

File: tests/app/dao/test_services_dao.py (hunk 5 above)
Comment (CrystalPea) [reply tests/app/dao/test_services_dao.py]: I added comments to this test so it's clearer what happens there :)

File: app/dao/services_dao.py [hunk 7]
```diff
@@ -72,6 +75,93 @@ def dao_count_live_services():
...
+        Service.go_live_user_id,
+        User.name,
+        User.email_address,
+        User.mobile_number,
+        Service.go_live_at,
+        Service.volume_sms,
+        Service.volume_email,
+        Service.volume_letter,
+        this_year_ft_billing.c.notification_type
+    ).all()
```
Comment (servingUpAces) [inline app/dao/services_dao.py]: Do we need to order the results? Maybe by go_live_at?

Comment (CrystalPea) [reply app/dao/services_dao.py]: That's a good idea 👍

**Pull request merged**